engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/time/AbstractInstantBasedTimeValidator.java

```diff
@@ -1,64 +0,0 @@
-/*
- * Hibernate Validator, declare and validate application constraints
- *
- * License: Apache License, Version 2.0
- * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
- */
-package org.hibernate.validator.internal.constraintvalidators.bv.time;
-
-import java.lang.annotation.Annotation;
-import java.lang.invoke.MethodHandles;
-import java.time.Clock;
-import java.time.Instant;
-
-import javax.validation.ClockProvider;
-import javax.validation.ConstraintValidator;
-import javax.validation.ConstraintValidatorContext;
-
-import org.hibernate.validator.internal.util.logging.Log;
-import org.hibernate.validator.internal.util.logging.LoggerFactory;
-
-/**
- * Base class for all time validators that use an {@link Instant} to be compared to the time reference.
- *
- * @author Alaa Nassef
- * @author Guillaume Smet
- */
-public abstract class AbstractInstantBasedTimeValidator<C extends Annotation, T> implements ConstraintValidator<C, T> {
-
-	private static final Log LOG = LoggerFactory.make( MethodHandles.lookup() );
-
-	@Override
-	public boolean isValid(T value, ConstraintValidatorContext context) {
-		// null values are valid
-		if ( value == null ) {
-			return true;
-		}
-
-		Clock reference;
-
-		try {
-			ClockProvider clockProvider = context.getClockProvider();
-			reference = clockProvider.getClock();
-		}
-		catch (Exception e) {
-			throw LOG.getUnableToGetCurrentTimeFromClockProvider( e );
-		}
-
-		int result = getInstant( value ).compareTo( reference.instant() );
```

Comment:
Hmmm, I'm a bit worried about replacing all the validator implementations in a micro release.

In an earlier comment, I pointed you to `minus()` and `plus()`. I think we should simply adjust the reference instant using the clock skew tolerance (in + or - depending if we validate a past or a future constraint). All the hierarchy is already ready for that. We just need to add an abstract method implemented in the abstract Past/Future implementations.

That would limit the changes to the minimum. And I don't think it's less clean.

WDYT?

Reply:
yeah.. I've used those methods when I added this method:
```java
protected Temporal adjustCurrentTime(Temporal nowValue) {
	return isFuture() ? nowValue.plus( tolerance ) : nowValue.minus( tolerance );
}
```
These changes that I've made are definitely not for the micro release :)
Let me try the "keeping current hierarchy" approach and I'll get back :)

Reply:
in a mean time here's one more of those "crazy" ideas :smile: 
What if instead of simple duration we pass something like:
```java
interface ToleranceSupplier {
	<A extends Annotation> Duration toleranceFor(Class timeClazz, Class<A> constraint);
}
```
by default we would use some
```java
public static class ConstantToleranceSupplier implements ToleranceSupplier{
	private final Duration tolerance;

	public ConstantToleranceSupplier(Duration tolerance) {
		this.tolerance = tolerance;
	}

	@Override
	public <A extends Annotation> Duration toleranceFor(Class timeClazz, Class<A> constraint) {
		return tolerance;
	}
}
```
which would create when a property is passed in xml config. And otherwise user would be able to somehow customize the tolerance based on a constraint and property type (for example +/- 1 day for `LocalDate` and +/- 10 sec for a `LocalDateTime`)